plugins/synaptics-rmi/fu-synaptics-rmi-device.c

```diff
@@ -0,0 +1,1098 @@
+
+static gboolean
+fu_synaptics_rmi_device_set_rma_page (FuSynapticsRmiDevice *self, guint8 page, GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+
+	/* same */
+	if (priv->page == page)
+		return TRUE;
+
+	/* write */
+	fu_byte_array_append_uint8 (req, page);
+	if (!fu_synaptics_rmi_device_write (self, RMI_DEVICE_PAGE_SELECT_REGISTER, req, error)) {
+		priv->page = -1;
+		return FALSE;
+	}
+	priv->page = page;
+	return TRUE;
+}
+
+gboolean
+fu_synaptics_rmi_device_reset (FuSynapticsRmiDevice *self, GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+
+	g_debug ("resetting...");
+	fu_byte_array_append_uint8 (req, RMI_F01_CMD_DEVICE_RESET);
+	if (!fu_synaptics_rmi_device_write (self, priv->f01->command_base, req, error))
+		return FALSE;
+	g_usleep (1000 * RMI_F01_DEFAULT_RESET_DELAY_MS);
+	g_debug ("reset completed");
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_device_scan_pdt (FuSynapticsRmiDevice *self, GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	guint interrupt_count = 0;
+
+	/* clear old list */
+	g_ptr_array_set_size (priv->functions, 0);
+
+	/* scan pages */
+	for (guint page = 0; page < RMI_DEVICE_MAX_PAGE; page++) {
+		guint page_start = RMI_DEVICE_PAGE_SIZE * page;
+		guint pdt_start = page_start + RMI_DEVICE_PAGE_SCAN_START;
+		guint pdt_end = page_start + RMI_DEVICE_PAGE_SCAN_END;
+
+		/* set page */
+		if (!fu_synaptics_rmi_device_set_rma_page (self, page, error))
+			return FALSE;
+
+		/* read out functions */
+		for (guint addr = pdt_start; addr >= pdt_end; addr -= RMI_DEVICE_PDT_ENTRY_SIZE) {
+			g_autofree FuSynapticsRmiFunction *func = NULL;
+			g_autoptr(GByteArray) res = NULL;
+			res = fu_synaptics_rmi_device_read (self, addr, RMI_DEVICE_PDT_ENTRY_SIZE, error);
+			if (res == NULL) {
+				g_prefix_error (error, "failed to read PDT entry @ 0x%04x: ", addr);
+				return FALSE;
+			}
+			func = fu_synaptics_rmi_function_parse (res, page_start, interrupt_count, error);
+			if (func == NULL)
+				return FALSE;
+			if (func->function_number == 0)
+				break;
+			interrupt_count += func->interrupt_source_count;
+			g_ptr_array_add (priv->functions, g_steal_pointer (&func));
+		}
+	}
+
+	/* see docs */
+	priv->num_interrupt_regs = (interrupt_count + 7) / 8;
+
+	/* for debug */
+	for (guint i = 0; i < priv->functions->len; i++) {
+		FuSynapticsRmiFunction *func = g_ptr_array_index (priv->functions, i);
+		g_debug ("PDT-%02u fn:0x%02x vr:%d sc:%d ms:0x%x "
+			 "db:0x%02x cb:0x%02x cm:0x%02x qb:0x%02x",
+			 i, func->function_number, func->function_version,
+			 func->interrupt_source_count,
+			 func->interrupt_mask,
+			 func->data_base,
+			 func->control_base, func->command_base,
+			 func->query_base);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_device_ioctl (FuSynapticsRmiDevice *self,
+			       gulong request,
```

Comment:
You only seem to be using `HIDIOCSFEATURE` other than the commented out code, maybe simplify arguments here?

Reply:
I suppose it's cleaner to pass in the datasz, indeed. Lets see if we need the HIDIOCGFEATURE and if not I can just nuke the abstraction.